---
title: "CHAS"
author: "Kitty Murphy, Alexi Nott, and Sarah Marzi"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
---

```{r setup, include=FALSE, dpi=400}
knitr::opts_chunk$set(echo = TRUE)

library(GenomicRanges)
library(data.table)
library(ggplot2)
library(wesanderson)
library(dplyr)
library(ggpubr)
```

```{r, echo=FALSE}
load("/Volumes/cm1118/home/CHAS/CHAS_rmd.RData")
```
# Citation

If you use the cell sorted H3K27ac data associated with this package 
then please cite the following paper: 
[Nott, et al. Brain cell type-specific enhancer-promoter interactome maps and disease-risk association. Science, 2019.](https://science.sciencemag.org/content/366/6469/1134.abstract)

# Background

Cell type identity is a major driver of epigenetic variation, making biological interpretation of bulk tissue epigenomes difficult. Here we present *CHAS* (cell type-specific histone acetylation score), an R package for inferring cell type-specific signatures in bulk brain H3K27ac profiles. *CHAS* annotates peaks identified in bulk brain studies of H3K27ac to cell type-specific signals in four major brain cell types, and based on signal intensities generates cell type-specific histone acetylation scores to act as a proxy for cell type proportion. *CHAS* was successfully validated in pseudo-bulk samples of known cell type proportions and applied to three neurodegenerative disorder epigenome-wide association studies conducted on bulk brain tissue. *CHAS* adds to the existing frameworks for cell type deconvolution in bulk brain tissue and we would expect to be able to extend the use of *CHAS* to other bulk tissue H3K27ac profiles. 

# *CHAS* workflow 

**1. Identification of cell type-specific peaks in bulk tissue H3K27ac profiles.**

*CHAS* annotates peaks identified in bulk tissue studies of H3K27ac to their cell type-specific signals by overlapping the bulk peaks with cell sorted H3K27ac peaks and identifying which of the bulk peaks are specific to a given cell type. For a bulk peak to be defined as cell type-specific two criteria must be met: (i) the bulk peak is annotated only to a single cell type; (ii) the bulk peak overlaps a predefined percentage of that cell typeâ€™s peak.

**2. Cell type-specific histone acetylation score generation.**

Using a counts per million matrix and the cell type-specific bulk H3K27ac peaks identified in step 1 of the workflow, *CHAS* generates scores by averaging the normalised signal intensity of a sample across all peaks specific to a given cell type, thereby deriving a proxy of the proportion of that cell type in the given bulk sample. As a constraint from peak-normalisation, the maximum signal intensity for any given peak and sample is 1 and the resulting score will lie between 0 and 1 for a given sample and cell type.

# Installing *CHAS* 

```
if (!require("devtools")) {
  install.packages("devtools")
}
devtools::install_github("neurogenomics/CHAS")
```

# Tutorial 

*CHAS* requires 3 inputs to run:

1. **Bulk tissue H3K27ac peaks in BED format (Column 1 - chrom, Column 2 - chromStart, Column 3 - ChromEnd, Column 4 - name)**. The columns can be named as you like but __the order must be as stated here.__ 
2. **Cell sorted H3K27ac peaks in BED format (Col 1 - chrom, Col 2 - chromStart, Col 3 - ChromEnd, Col 4 - name)**. The columns can be named as you like but __the order must be as stated here.__ 
3. **Counts per million mapped reads (cpm) matrix**, with rows labelling peaks and columns labelling samples.

**Note:** If you are using the cell sorted H3K27ac peaks provided in this package, **column 1 of your bulk tissue H3K27ac peaks file must be in this format - chrN**, where N is the chromosome name e.g. 1 or X. 

In this tutorial we will use entorhinal cortex H3K27ac peaks from [Marzi et al. 2018](https://www.nature.com/articles/s41593-018-0253-7), cell sorted H3K27ac peaks from [Nott et al. 2019](https://science.sciencemag.org/content/366/6469/1134.long), and a counts per million matrix constructed using the entorhinal cortex H3K27ac data from [Marzi et al. 2018](https://www.nature.com/articles/s41593-018-0253-7). 

### 1. Load the example data

Bulk brain peaks: 
```
data(EntorhinalCortex_AD_H3K27ac_peaks)
```

Counts per million matrix: 
```
data(EntorhinalCortex_AD_H3K27ac_counts)
```

### 2. Load the input data

<p>The cell sorted H3K27ac peaks are available in two human genome reference builds: hg19 and hg38.<br> 
To load the **hg19** peaks:</p>

```
data(astro_H3K27ac_hg19)
data(mgl_H3K27ac_hg19)
data(neu_H3K27ac_hg19)
data(olig_H3K27ac_hg19)
```

To load the **hg38** peaks:

```
data(astro_H3K27ac_hg38)
data(mgl_H3K27ac_hg38)
data(neu_H3K27ac_hg38)
data(olig_H3K27ac_hg38)
```


### 3. Generate list containing data frames of all cell sorted H3K27ac peaks

This example uses the cell sorted peaks in **hg38** build. 

```
celltypePeaks <- list(Astrocyte = astro_H3K27ac_hg38, Microglia = mgl_H3K27ac_hg38, Neuron = neu_H3K27ac_hg38, Oligodendrocyte = olig_H3K27ac_hg38)
```

### 4. Annotate bulk H3K27ac peaks and identify which are cell type-specific

```
celltype_specific_peaks <- CelltypeSpecificPeaks(EntorhinalCortex_AD_H3K27ac_peaks, celltypePeaks, 0.5) 
```

*CelltypeSpecificPeaks()* takes as input the bulk peaks, the list of cell sorted peaks, and a value from 0-1 specifying the percentage overlap required for a bulk peak to be considered cell type-specific, and outputs a list of two data frames. The first data frame contains all bulk peaks that are specific to a cell type and the second data frame contains all the bulk peaks annotated either to a cell type, 'multiple' cell types, or 'other'. 

### 5. Generate cell type-specific histone acetylation scores
```
celltype_scores <- CelltypeScore(EntorhinalCortex_AD_H3K27ac_counts, celltype_specific_peaks) 
```

*CelltypeScore()* uses a counts per million matrix and the output of the function *CelltypeSpecificPeaks()* to generate cell type-specific H3K27ac scores for each sample in the counts matrix. 

### 6. Plot the cell type proportions in the bulk H3K27ac peaks. 

``` {r, fig.width=5, fig.height=3.5}
plot_celltype_props(celltype_specific_peaks, exampleData=TRUE)
```

*plot_celltype_props()* uses the annotated bulk H3K27ac peaks from *CelltypeSpecificPeaks()* to generate a stacked bar plot of the proportions of each cell type in the bulk H3K27ac peak set. Set exampleData=TRUE if you have used the cell sorted data provided within *CHAS*.

### 7. Plot the cell type H3K27ac scores between two different groups.

``` {r, fig.width=4.5, fig.height=3.5}
plot_celltype_scores(celltype_scores, AD_pheno)
```


*plot_celltype_scores()* requires the cell type-specific scores generated using *CelltypeScore()* and a data frame containing a column named Sample with the sample ID's that match those labelling the columns in the counts per million matrix and another column named Group containing the group to which each sample belongs to e.g. case or control. The output is a violin plot of the cell type-specific H3K27ac scores in the different groups. 
